src/monodroid/jni/android-system.cc

```diff
@@ -731,89 +732,68 @@ AndroidSystem::get_gref_gc_threshold ()
 
-	if ((k = entry [0]) && *k &&
-			(v = entry [1]) && *v) {
-		if (islower (k [0])) {
-			add_system_property (k, v);
-		} else {
-			setenv (k, v, 1);
-		}
-	}
+	if (islower (k [0])) {
+		if (k [0] == 'm') {
+			if (strcmp (k, "mono.aot") == 0) {
+				knownEnvVars.MonoAOT = strdup (v);
+				return;
+			}
 
-	utils.monodroid_strfreev (entry);
-}
+			if (strcmp (k, "mono.llvm") == 0) {
+				knownEnvVars.MonoLLVM = true;
+				return;
+			}
+		}
 
-void
-AndroidSystem::setup_environment_from_file (const char *apk, int index, int apk_count, void *user_data)
-{
-	unzFile file;
-	if ((file = unzOpen (apk)) == nullptr)
+		add_system_property (k, v);
 		return;
+	}
 
-	if (unzLocateFile (file, "environment", 0) == UNZ_OK) {
-		unz_file_info info;
-
-		if (unzGetCurrentFileInfo (file, &info, nullptr, 0, nullptr, 0, nullptr, 0) == UNZ_OK &&
-				unzOpenCurrentFile (file) == UNZ_OK) {
-			char *contents = new char [info.uncompressed_size+1];
-			if (contents != NULL &&
-					unzReadCurrentFile (file, contents, info.uncompressed_size) > 0) {
-
-				int i;
-				char *line = contents;
-				contents [info.uncompressed_size] = '\0';
-
-				for (i = 0; i < info.uncompressed_size; ++i) {
-					if (contents [i] != '\n')
-						continue;
-
-					contents [i] = '\0';
-					setup_environment_from_line (line);
-					line = &contents [i+1];
-				}
-
-				if (line < (contents + info.uncompressed_size))
-					setup_environment_from_line (line);
-
-				free (contents);
-			}
-
-			unzCloseCurrentFile (file);
+	if (k [0] == '_') {
+		if (strcmp (k, "__XA_DSO_IN_APK") == 0) {
+			knownEnvVars.DSOInApk = true;
+			return;
 		}
 	}
 
-	unzClose (file);
+	setenv (k, v, 1);
 }
 
 void
-AndroidSystem::for_each_apk (JNIEnv *env, jobjectArray runtimeApks, void (AndroidSystem::*handler) (const char *apk, int index, int apk_count, void *user_data), void *user_data)
+AndroidSystem::setup_environment (JNIEnv *env, jobjectArray environmentVariables)
 {
-	int i;
-	jsize apksLength = env->GetArrayLength (runtimeApks);
-	for (i = 0; i < apksLength; ++i) {
-		jstring e       = reinterpret_cast<jstring> (env->GetObjectArrayElement (runtimeApks, i));
-		const char *apk = env->GetStringUTFChars (e, nullptr);
-
+	jsize envvarsLength = env->GetArrayLength (environmentVariables);
+	if (envvarsLength == 0)
+		return;
 
-		(this->*handler) (apk, i, apksLength, user_data);
-		env->ReleaseStringUTFChars (e, apk);
+	jstring_wrapper name (env), value (env);
+	for (jsize i = 0; i < envvarsLength; i += 2) {
+		name = reinterpret_cast<jstring> (env->GetObjectArrayElement (environmentVariables, i));
+		value = reinterpret_cast<jstring> (env->GetObjectArrayElement (environmentVariables, i + 1));
+		setup_environment (name, value);
 	}
 }
 
 void
-AndroidSystem::setup_environment (JNIEnv *env, jobjectArray runtimeApks)
+AndroidSystem::for_each_apk (JNIEnv *env, jstring_array_wrapper &runtimeApks, void (AndroidSystem::*handler) (const char *apk, int index, int apk_count, void *user_data), void *user_data)
```

Comment:
Could `handler` be...whatever the newfangled C++ delegate/closure construct is?  Or would that have a noticeable perf impact?

Reply:
You mean a lambda? It could, but then we'd have to use some of STL (`std::function`). Alas, we already make use of STL (`std::atomic` and `std::mutex`) but I hope to replace them with something custom in the future.

Reply:
Yes, lambda is what i was thinking, and I didn't realize it required the STL/libstdc++.  Doh!

I just hoped that lambdas could be converted to something without dependencies...